Add dataset release utils

tuanchien opened the pull request:
- Fixes deprecated notation in dockercompose for observatory: "postgres" -> "postgresql" in the uri.
- Adds basic utility functions to assist with adding dataset releases to the api.
- Add basic Airflow tasks to assist with adding and updating dataset release records in the api from workflows.
- Changes stream telescope behaviour to not raise skip exceptions in bq tasks but instead to return success, and just do nothing.

Of note:
- 'doi workflow' in the academic observatory uses `find_dataset_id_from_name` to find dataset ids. This needs the dataset names to be in some standardised form to be found. See: https://github.com/The-Academic-Observatory/observatory-platform/blob/2f60be875da4ecf0c32ed01f79f78daabd3a3a98/observatory-platform/observatory/platform/utils/release_utils.py#L100
- jstor and google books telescopes use `find_dataset_id_from_address` which gets dataset ids from dataset storage records instead.  With the way the api backend is currently designed, this generates slighly more traffic, but the address field is maybe more stable of a thing to use to find the dataset_id. See: https://github.com/The-Academic-Observatory/observatory-platform/blob/2f60be875da4ecf0c32ed01f79f78daabd3a3a98/observatory-platform/observatory/platform/utils/release_utils.py#L116

The two implementations are kind of there to as experiments to highlight two different approaches. Would be good for you to comment on what pattern you would prefer to use.

[Update]
Changed `doi workflow` to use dataset storage address based dataset id discovery instead of dataset name based.

aroelo reviewed (commented):
Hey @tuanchien, thank you for those changes.

I have had a global look at them as well as in the project specific repositories.
I'm just wondering if it would be possible to add some more documentation or explanations on the added code?
Kind of like the API use cases here https://observatory-platform--540.org.readthedocs.build/en/540/api/api_use_cases.html, but maybe with the addition of what the definition for each of the following concepts is and the relationships to each other:
Dataset, DatasetRelease, DatasetStorage. 
I think it would be useful documentation in general and it will make it easier to review the changes.
We should probably have something similar for the other API concepts like Organisation, Telescope and TelescopeType, but that is not in scope for this PR.

As far as I can see so far the changes only add the creation of a dataset release, but is not using a created DatasetRelease anywhere in the workflows yet? I may have missed this somewhere though.

I'm also curious about the differences between `update_dataset_release` and `add_dataset_release`.
It seems like for streamtelescopes, where there is a partitioned table, the dataset release is updated instead of a new release created.
What would be the idea or definition behind a 'datasetrelease'?
I expected that for streamtelescopes each partition would also be a new dataset release, but this might not make the most sense, curious to hear the reasoning or use cases behind this decision.

In terms of the implementation, considering that a dataset release will be created for every workflow, what do you think about adding the `add_dataset_release_task` function as a method to the base Workflow class?
Similar to e.g. 'check_dependencies' https://github.com/The-Academic-Observatory/observatory-platform/blob/develop/observatory-platform/observatory/platform/workflows/workflow.py#L385-L402.
The kwargs that it needs for this function like download_task_id or schema_version could then be set as class attributes as part of the init() method.

I'm thinking this might be useful so it becomes easier to add the task with `self.add_task(self.add_dataset_release)`, and it seems more in line with the other tasks of the template.

jdddog commented on observatory-platform/observatory/platform/utils/release_utils.py:
```diff
@@ -0,0 +1,449 @@
+
+    dataset = get_dataset(telescope_type_id=telescope_type_id, org_name=org_name)
+    return dataset.id
+
+
+def find_dataset_id_from_name(*, name: str, datasets: List["Dataset"]) -> int:  # noqa: F821
+    """Get the dataset id for a given dataset name, from a list of Datasets.
+    If no dataset is found, it will throw an exception.
+
+    :param name: Name of the dataset to search for.
+    :param datasets: List of datasets to search.
+    :return: Dataset id.
+    """
+
+    for dataset in datasets:
+        if dataset.name == name:
+            return dataset.id
+
+    raise Exception(f"Dataset {name} not found")
+
+
+def find_dataset_id_from_address(*, address: str, datasets: List["Dataset"]) -> int:  # noqa: F821
+    """Get the dataset id for a given dataset storage address, from a list of Datasets.
+    If no dataset is found, it will throw an exception.
+
+    :param address: Storage address of the dataset to search for.
+    :param datasets: List of datasets to search.
+    :return: Dataset id.
+    """
+
+    limit = 1000
+
+    for dataset in datasets:
+        api = make_observatory_api()
+        storages = api.get_dataset_storages(dataset_id=dataset.id, limit=limit)
+        for storage in storages:
+            if storage.address == address:
+                return dataset.id
+
+    raise Exception(f"Dataset with address {address} not found")
```
These functions seem like they are going to create brittle code because they are determining the dataset_id based on names and dataset addresses.

tuanchien commented on observatory-platform/observatory/platform/utils/release_utils.py:
```diff
@@ -0,0 +1,449 @@
+import datetime
+
+
+def address_to_gcp_fields(address: str) -> Tuple[str, str, str]:
+    """Take an address in the DatasetStorage field, and return the project ID, dataset ID, table ID.
+
+    :param address: Address string in the DatasetStorage field.
+    :return GCP project, dataset, table.
+    """
+
+    return tuple(address.split("."))
```
It's currently unused since that same information is duplicated in several different places, and there is also very tight coupling with gcp. However, in the event that the information duplication is removed, or telescopes start to be written that are more cloud platform agnostic, the service and address fields can be used with parsers like this to retrieve the table URI information.

tuanchien commented:
Thanks for the feedback @aroelo @jdddog.  Some of the 2cents/responses have been added to to the "Observatory Database & API Review" google doc just so the main points are kept together for the Thursday discussion.

tuanchien commented:
Updates:
- TableType, DatasetType, Dataset, DatasetStorage, DatasetRelease changes as per discussion doc. Exclusions: Telescope not renamed to Workflow, Organisation not made optional.
- Schema version info also removed from DatasetRelease.
- add_new_dataset_releases task added to Workflow class. All workflow types will use this to create new dataset releases when there are new ingestion runs.  No longer needs additional arguments. It will try to discover all datasets for a given workflow id, and create a new datasetrelease for each dataset.
- update_dataset_release_task and related functions are removed.
- get_dataset removed. Now just using get_datasets.
- project_id is added as a parameter to a lot of bq functions in workflow_utils to give granular control.  In the past the project id of the bq client was assumed to always be the target project_id. However, given some workflow test datasets are only available on bq and not local, you need granular control over project_id to access those tables.
- Added get_filename_from_http_header to discover filename from http header when downloading a file from url.
- Sorting bq info by transfer blob names so that loading order is more deterministic.
- disable catchup in stream telescope
- refactor get_release_info in stream telescope to make use of datasetrelease info and Airflow data interval. Change data period from [a,b] to [a,b) to be more consistent with data intervals.
- changed some dates (usually shifted forward by a day) in unit tests in response to discovered quirky behaviour of how data intervals are calculated by create_dagrun.
- update stream telescope generator template to make tests pass

TODO:
- documentation updates

tuanchien commented:
Updates:
- Telescope record changed to Worfklow
- gcp prefix removed from fields in Organisation
- Update documentation with api use cases and basic record query and updates.

jdddog reviewed (changes requested):
Thanks Tuan, the changes look good, just a few minor comments.

tuanchien commented on observatory-api/observatory/api/server/orm.py:
```diff
@@ -131,161 +131,228 @@ class Organisation(Base):
     id = Column(Integer, primary_key=True)
     name = Column(String(250))
     extra = Column(JSON())
+    tags = Column(String(250), nullable=True)
```
If you specify it as a JSON() it expects a dictionary during parsing, so it won't like it if a list is passed in instead.

jdddog reviewed (approved):
Thanks for the changes Tuan, it looks good to me.
